interfaces/builtin/service_watchdog.go

```diff
@@ -0,0 +1,73 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2018 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+import (
+	"os"
+	"strings"
+
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/interfaces/apparmor"
+)
+
+const serviceWatchdogSummary = `allows use of systemd service watchdog`
+
+const serviceWatchdogBaseDeclarationSlots = `
+  service-watchdog:
+    allow-installation:
+      slot-snap-type:
+        - core
+    deny-auto-connection: true
+`
+
+const serviceWatchdogConnectedPlugAppArmorTemplate = `
+# Allow sending notification messages to systemd through the notify socket
+{{notify-socket}} w,
+`
+
+type serviceWatchdogInterface struct {
+	commonInterface
+}
+
+var osGetenv = os.Getenv
+
+func (iface *serviceWatchdogInterface) AppArmorConnectedPlug(spec *apparmor.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	notifySocket := osGetenv("NOTIFY_SOCKET")
```

Comment:
If we are going to pull something from the environ and shove it into an AppArmor rule, then we need to make sure the path is absolute and doesn't contain any AARE characters that might cause the rule to misparse. Something along the lines of https://github.com/snapcore/snapd/blob/master/osutil/overlay.go#L52 would do. Refactoring the regex or the whole check would be good for this and overlay.go to both use.